Include all Teleport CAs in the rotation guide

ptgott opened the pull request:
Closes #37125
Closes #44398

This change clarifies the differences between Teleport CAs in the CA Rotation guide. Since the Database CA Rotation guide is almost identical to the original CA Rotation guide, this change merges the two guides.

To include relevant information for each CA, this change adds a step to help the user choose a CA to rotate, and includes specific information for each CA in that step.

Another possibility would be to spread information for each CA across the instructions for each transition phase. However, since each CA requires a different set of checks/updates in different phases, it makes sense to make the instructions for each CA more free-flowing while restricting them to a single, early section.

Most information about checking the status of a rotation in the original guide had to do with the `host` CA, and this change moves that information into an H3 related to that CA.

ptgott committed:
Include all Teleport CAs in the rotation guide

Closes #37125

This change clarifies the differences between Teleport CAs in the CA
Rotation guide. Since the Database CA Rotation guide is almost identical
to the original CA Rotation guide, this change merges the two guides.

To include relevant information for each CA, this change adds a step to
help the user choose a CA to rotate, and includes specific information for
each CA in that step.

Another possibility would be to spread information for each CA across
the instructions for each transition phase. However, since each CA
requires a different set of checks/updates in different phases, it makes
sense to make the instructions for each CA more free-flowing while
restricting them to a single, early section.

Most information about checking the status of a rotation in the original
guide had to do with the `host` CA, and this change moves that
information into an H3 related to that CA.

ptgott commented:
Hi @GavinFrazar, it'd be great to get a review from you if you have time, since you wrote the original database CA rotation guide and have done recent work on our CAs. Thanks!

GavinFrazar commented:
> Hi @GavinFrazar, it'd be great to get a review from you if you have time, since you wrote the original database CA rotation guide and have done recent work on our CAs. Thanks!

Awesome! I was about to propose essentially the same thing as this, but you beat me by 3 weeks lol

I'll try to take a look today as I'll be on vacation all next week

zmb3 commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,448 @@
 ---
 
+Various components of a Teleport cluster authenticate to one another using
+either X.509 or SSH certificates. To verify certificates, Teleport maintains
+several certificate authorities. You can rotate Teleport CAs to prevent
```
I would say that Teleport uses the certificate authorities to _issue certificates_. Verification against the CA is done by the target resources (Windows desktop, database, SSH node, etc)

zmb3 commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,448 @@
-  issued by the old certificate authority.
-- `rollback`: The rotation was aborted and is rolling back to the old
-  certificate authority.
+|CA type|Traffic it authenticates|
```
```suggestion
|CA type|Certificates it issues|
```

Perhaps?

zmb3 commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,448 @@
+### `host`
 
-There are two kinds of certificate rotations:
+The `host` CA authenticates traffic between Teleport agents and Proxy Service
+instances and the Teleport Auth Service.
```
```suggestion
The `host` CA issues certificates for Teleport cluster components (agents
 and Proxy Service
instances and the Teleport Auth Service).
```

zmb3 reviewed (commented):
My comments apply throughout the whole doc and may be a matter of preference, but I would frame the doc around what certificates each CA issues instead of around what traffic it authenticates.

(Technically, CAs don't really authenticate anything. They issue certs. The entities communicating with those certs are responsible for authenticating each other.)

ptgott committed:
Respond to zmb3 feedback

- Edit language that suggests a CA authenticates with a service.
- Add a "Verified by" column to the CA table and indicate which party
  verifies certificates issues by each CA.
- Indicate that the `host` CA issues certificates for the Auth Service.

GavinFrazar commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,462 @@
 
-Initiate the manual rotation of host certificate authorities:
+|CA type|Certificate subjects|Verified by|
+|---|---|
```
table rendering fix:
```suggestion
|---|---|---|
```

GavinFrazar replied:
I think we should drop the "verified by" column.

Accurately describing and enumerating the things that verify certs issued by a given CA is too much info.

GavinFrazar commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,462 @@
+1. `init`: A new certificate authority is issued, but not used.
+1. `update_clients`: The Teleport Auth Service uses the new CA to sign
+   certificates but continues to trust certificates signed by the original CA.
+1. `update_servers`: Teleport servers (Agents, Auth Service, and Proxy Service
```
```suggestion
1. `update_servers`: Teleport cluster components (Agents, Auth Service, and Proxy Service
```

GavinFrazar commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,462 @@
+
+### `update_servers`
+
+Initiate the `update_servers` phase. In this phase, Teleport servers (Agents,
```
```suggestion
Initiate the `update_servers` phase. In this phase, Teleport cluster components (Agents,
```

GavinFrazar commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,462 @@
+**should not** carry out a semi-automatic rotation if you need to export a CA to
+your infrastructure.
+
+Complete a rotation in  manual mode first to understand all the edge-cases and
```
```suggestion
Complete a rotation in manual mode first to understand all the edge-cases and
```

GavinFrazar commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,462 @@
+the current phase before the grace period elapses.
+
+Teleport does not check the status of any infrastructure that relies on the CA,
+meaning that you can lose connectivity if things go wrong. As a result, you
+**should not** carry out a semi-automatic rotation if you need to export a CA to
+your infrastructure.
```
@zmb3 what do you think about a code change to just reject automatic rotation requests for several of our CAs? For example, I can't think of a legit use case for doing automatic rotation with `db` or `db_client` CAs, because you have to reissue all of your database certificates or reconfigure them to trust the new CA cert. An automatic rotation will just break access to self-hosted databases like a ticking time bomb.

As far as I know, host and user CA automatic rotation is fine (maybe not user if there are desktops configured?). Not sure about other ones.

GavinFrazar commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,462 @@
+
+The Teleport Database Service presents a certificate signed by the `db_client`
+CA when communicating with a self-hosted database, which an admin configures to
+trust certificates issued by the CA. Self-hosted databases present a certificate
+signed by the `db` CA, which the Database Service uses to verify that a database
+server is a genuine Teleport-protected resource.
```
```suggestion
trust certificates issued by the CA. Self-hosted databases can be issued a certificate by the `db` CA, which the Database Service uses by default to verify that a database
server is a genuine Teleport-protected resource.
```

it's optional to use a `db` CA cert - they can their own custom CA to issue a cert for their databases, but they must configure the db service to trust their custom CA

GavinFrazar reviewed (approved):
I have a few suggestions but nothing worth blocking the PR for.

By the way you need to remove all links to the deleted db-ca-rotation.mdx file (one in operations.mdx, the other in db-ca-migrations.mdx)

ptgott committed:
Respond to GavinFrazar feedback

- Remove the "verified" column from the initial table.
- Change "servers" to "cluster components".
- Clarify that database servers can use a custom CA.

ptgott commented:
@zmb3 Putting this on your radar for when you have some time to review. Thanks!

ptgott commented:
@zmb3 Following up on this one for when you have time for another look. Thanks!

zmb3 commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,466 @@
+   [`host` CA](#host).
+   1. `standby`: No rotation in progress. All operations have completed.
+
+Before the final `standby` phase, you an also put the rotation in the `rollback`
```
```suggestion
Before the final `standby` phase, you can also put the rotation in the `rollback`
```

zmb3 commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,466 @@
 
-**Start the rotation**
+We recommend rotating a single CA at a time in order to reduce complexity. The
+exception is rotating the `db` and `db_client` CAs. 
```
```suggestion
exception is rotating the `db` and `db_client` CAs. These two CAs must
be rotated together.
```

Not sure if my language here is exactly right, but we should explain what the exception we refer to is.

zmb3 commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,466 @@
+|CA type|Certificate subjects|
+|---|---|
+|[`host`](#host)|Teleport Agents. Auth Service and Proxy Service instances.|
+|[`user`](#user)|Teleport users. Teleport SSH Service and Windows Desktop Service instances.|
```
The _subjects_ in the certificates issued by these CAs are still Teleport users, not Teleport services (those still use the host CA).

zmb3 commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,466 @@
+|[`host`](#host)|Teleport Agents. Auth Service and Proxy Service instances.|
+|[`user`](#user)|Teleport users. Teleport SSH Service and Windows Desktop Service instances.|
+|[`db`](#db-and-db_client)|Self-hosted databases protected by Teleport (users must distribute certificates to databases).|
+|[`db_client`](#db-and-db_client)|The Teleport Database Service.|
```
Is this correct? The Teleport Database Service is a server, right? Seems odd that a server cert would be issued by a client CA?

GavinFrazar replied:
This is correct, the Teleport Database Service is issued a client cert by the `db_client` CA at connection time, which it then uses to authenticate itself to the database as a client.

zmb3 commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,466 @@
+|[`db`](#db-and-db_client)|Self-hosted databases protected by Teleport (users must distribute certificates to databases).|
+|[`db_client`](#db-and-db_client)|The Teleport Database Service.|
+|[`openssh`](#openssh)|OpenSSH servers enrolled in your Teleport cluster.|
+|[`jwt`](#jwt)|Teleport users|
```
Perhaps?

```suggestion
|[`jwt`](#jwt)|Teleport users accessing web applications|
```

zmb3 commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,466 @@
+|---------------------|------------------|
+| Application Service | `app_server`     |
+| Database Service    | `db_server`      |
+| Kubernetes Service  | `kube_server`    |
```
```suggestion
| Kubernetes Service  | `kube_server`    |
| Windows Desktop Service | `windows_desktop_service` |
```

Assuming rotation status is included for w_d_s

zmb3 commented on docs/pages/management/operations/ca-rotation.mdx:
````diff
@@ -1,261 +1,466 @@
-  "phase": "update_servers"
-}
-```
+If you are joining Teleport processes to a cluster via the Teleport Auth Service
+using a [join token](../../agents/join-services-to-your-cluster/join-token.mdx),
````
```suggestion
If you are joining Teleport processes to a cluster by dialing the Teleport Auth Service directly,
```

Not sure what the join token has to do with this. All joining requires a join token.

zmb3 commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,466 @@
+### `user`
+
+The `user` CA issues a certificate when a user authenticates to Teleport. It
+also signs client certificates for the Windows Desktop Service and Teleport SSH
+Service. Teleport-protected servers and Windows desktops use these certificates
```
```suggestion
also signs client certificates for users connecting to Windows desktops and Teleport SSH servers.
Teleport-protected servers and Windows desktops use these certificates
```

zmb3 commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,466 @@
+certificate from the new CA, otherwise Teleport client commands fail.
+
+If you have registered Windows desktops with Teleport, [follow the
+guide](../../desktop-access/active-directory-manual.mdx) to exporting the
```
```suggestion
guide](../../desktop-access/active-directory-manual.mdx) to export the
```

zmb3 commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,466 @@
+
+### `oidc_idp`
+
+The `oidc_idp` CA signs messages sent by the Teleport OIDC IdP integration. AWS
```
```suggestion
The `oidc_idp` CA signs messages sent by the Teleport OIDC AWS integration. AWS
```

GavinFrazar replied:
We will expand the OIDC integration to other clouds, I think Azure is already in the works.

I would stick with the generic IdP language here, maybe even scrap the AWS specific language and speak generically about "the cloud" or "the relying party"?

zmb3 commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,466 @@
+
+The `oidc_idp` CA signs messages sent by the Teleport OIDC IdP integration. AWS
+verifies these messages to authenticate your Teleport account for features like
+EICE EC2 Access, External Audit Storage, Auto-Discovery, and AWS Sync for Access
```
I think we removed EICE support from Teleport, didn't we?

GavinFrazar replied:
yes, all the code is still there but we removed the tile from the web ui

zmb3 commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -1,261 +1,466 @@
+certificate authority:
+
+- From the `/.well-known/jwks-oidc` path of the Teleport Proxy Service Web API.
+- From an Amazon S3 bucket.
```
We are in the process of removing the S3 bucket, so I think we can remove this.

zmb3 reviewed (commented):
Sorry for losing track of this one!

ptgott committed:
Respond to zmb3 feedback

- Accuracy and readability improvements.
- Remove the JWKS OIDC discussion in relation to S3.

GavinFrazar commented on docs/pages/management/operations/ca-rotation.mdx:
```diff
@@ -240,47 +240,16 @@ metadata file and make it available to your service provider.
-- From the `/.well-known/jwks-oidc` path of the Teleport Proxy Service Web API.
-- From an Amazon S3 bucket.
+The Teleport Proxy Service serves the public key for the Teleport `oidc_idp`
+certificate authority from the `/.well-known/jwks-oidc` path of the Web API.
```
nit: I would mention `/.well-known/openid-configuration` as well, which includes the path to the jwks uri, e.g. mine looks like:
```json
{
  "issuer": "https://alpha.devteleport.com",
  "jwks_uri": "https://alpha.devteleport.com/.well-known/jwks-oidc",
  "claims": [
    "iss",
    "sub",
    "obo",
    "aud",
    "jti",
    "iat",
    "exp",
    "nbf"
  ],
  "id_token_signing_alg_values_supported": [
    "RS256"
  ],
  "response_types_supported": [
    "id_token"
  ],
  "scopes_supported": [
    "openid"
  ],
  "subject_types_supported": [
    "public",
    "pair-wise"
  ]
}
```

ptgott committed:
Respond to GavinFrazar feedback

Mention the `/.well-known/openid-configuration` path.

ptgott commented:
@zmb3 Following up to see if my latest changes satisfy your feedback. Thanks!

ptgott commented:
Hi @zmb3, checking again to see if I have resolved your feedback. Thanks!
